File: JBTheatreToolsWin/Core/VersionCompare.cs
namespace JBTheatreTools;

/// <summary>Lenient, overflow-safe version-string comparison (audit F13). Tolerates a leading "v" and
/// differing component counts; parses each numeric segment into a <c>long</c> with saturation, so a long
/// numeric tag (e.g. "v20260906123456") can't wrap a 32-bit int and compare as OLDER than the installed
/// build — which would silently hide an update. Shared by the WinForms <c>Versions</c> facade so there's
/// one comparator, unit-tested here.</summary>
public static class VersionCompare
{
    public static string Norm(string s)
    {
        s = s.Trim();
        return s.StartsWith('v') || s.StartsWith('V') ? s[1..] : s;
    }

    public static bool Equal(string a, string b) => Norm(a) == Norm(b);

    /// <summary>True if `a` is a strictly newer version than `b`.</summary>
    public static bool IsNewer(string a, string b) => Compare(a, b) > 0;

    /// <summary>Semver-aware ordering: the numeric core compares component-wise (missing parts = 0); a
    /// pre-release (<c>1.2.0-dev.3</c>, <c>1.2.0-rc1</c>) sorts BEFORE its release, and pre-release identifiers
    /// compare semver-style (numbers numerically, numbers before words, a longer list after its prefix) — so
    /// <c>0.1.0-dev.1 &lt; 0.1.0-dev.2 &lt; 0.1.0</c>. A tag that doesn't start with a digit (Convert's
    /// <c>build-20260912</c>) keeps the lenient digit-run parse and never counts as a pre-release.</summary>
    public static int Compare(string a, string b)
    {
        var (ca, pa) = SplitPre(a);
        var (cb, pb) = SplitPre(b);
        long[] x = Parts(ca), y = Parts(cb);
        for (int i = 0; i < Math.Max(x.Length, y.Length); i++)
        {
            long u = i < x.Length ? x[i] : 0;
            long v = i < y.Length ? y[i] : 0;
            if (u != v) return u.CompareTo(v);
        }
        if (pa == null && pb == null) return 0;
        if (pa == null) return 1;
        if (pb == null) return -1;
        for (int i = 0; i < Math.Min(pa.Length, pb.Length); i++)
        {
            int c = CompareIdentifier(pa[i], pb[i]);
            if (c != 0) return c;
        }
        return pa.Length.CompareTo(pb.Length);
    }

    /// <summary>A tag as the launchers print it: "v" + the normalised version for a numeric tag ("1.2.0" and
    /// "v1.2.0" both → "v1.2.0"), the tag itself otherwise (Convert's "build-20260912").</summary>
    public static string Display(string tag)
    {
        var n = Norm(tag);
        return n.Length > 0 && char.IsDigit(n[0]) ? "v" + n : tag.Trim();
    }

    /// <summary>True for development pre-releases (<c>vX.Y.Z-dev.N</c>).</summary>
    public static bool IsDev(string tag) => Norm(tag).Contains("-dev.", StringComparison.OrdinalIgnoreCase);

    /// <summary>The release to treat as "latest". Dev pre-releases are candidates ONLY with the Dev channel on
    /// (then: highest semver among stable + dev); off, they're invisible — even for an app whose only releases
    /// are dev builds. Otherwise the highest stable, falling back to the highest non-dev pre-release.</summary>
    public static T? PickLatest<T>(IEnumerable<T> releases, Func<T, string> tag, Func<T, bool> isPre, bool devChannel)
        where T : class
    {
        var list = releases.ToList();
        var nonDev = list.Where(r => !IsDev(tag(r))).ToList();
        var stable = nonDev.Where(r => !isPre(r)).ToList();
        var pool = devChannel ? stable.Concat(list.Where(r => IsDev(tag(r)))).ToList() : stable;
        if (pool.Count == 0) pool = nonDev;
        T? best = null;
        foreach (var r in pool)
            if (best == null || Compare(tag(r), tag(best)) > 0) best = r;
        return best;
    }

    private static (string Core, string[]? Pre) SplitPre(string s)
    {
        var n = Norm(s);
        int dash = n.IndexOf('-');
        if (n.Length == 0 || !char.IsDigit(n[0]) || dash < 0) return (n, null);
        return (n[..dash], n[(dash + 1)..].Split('.'));
    }

    private static int CompareIdentifier(string a, string b)
    {
        bool na = long.TryParse(a, out var x), nb = long.TryParse(b, out var y);
        if (na && nb) return x.CompareTo(y);
        if (na) return -1;
        if (nb) return 1;
        return string.CompareOrdinal(a, b);
    }

    private static long[] Parts(string s) => Norm(s).Split('.').Select(p =>
    {
        // First contiguous digit run in the segment: skip any leading non-digits, take the digits, stop at
        // the next non-digit. This handles date-style tags like "build-20260912" (→ 20260912) that a rolling
        // app such as Convert uses, while staying identical for ordinary semver segments ("2", "0-rc1" → 0).
        long n = 0;
        bool started = false;
        foreach (var c in p)
        {
            if (char.IsDigit(c))
            {
                started = true;
                n = n > (long.MaxValue - 9) / 10 ? long.MaxValue : n * 10 + (c - '0');   // saturate, never wrap
            }
            else if (started) break;
        }
        return n;
    }).ToArray();
}

File: JBTheatreToolsAndroid/app/src/main/java/com/jamesbreedon/jbtheatretools/core/VersionCompare.kt
package com.jamesbreedon.jbtheatretools.core

/**
 * Lenient, overflow-safe version-string comparison — the Kotlin twin of the Windows launcher's
 * `VersionCompare` and the macOS launcher's comparator.
 *
 * Tolerates a leading "v" and differing component counts, and parses each numeric segment into a
 * [Long] with saturation, so a long numeric tag (a date tag such as Convert's `build-20260912`)
 * can't wrap and compare as OLDER than the installed build — which would silently hide an update.
 */
object VersionCompare {
    fun norm(s: String): String {
        val t = s.trim()
        return if (t.startsWith("v") || t.startsWith("V")) t.substring(1) else t
    }

    fun equal(a: String, b: String): Boolean = norm(a) == norm(b)

    /** True if `a` is a strictly newer version than `b`. */
    fun isNewer(a: String, b: String): Boolean = compare(a, b) > 0

    /**
     * Semver-aware ordering. The numeric core compares component-wise (missing parts = 0); a pre-release
     * (`1.2.0-dev.3`, `1.2.0-rc1`) sorts BEFORE its release (`1.2.0`), and pre-release identifiers compare
     * semver-style (numbers numerically, numbers before words, a longer list after its prefix) — so
     * `0.1.0-dev.1 < 0.1.0-dev.2 < 0.1.0`. A tag that doesn't start with a digit (Convert's
     * `build-20260912`) keeps the lenient digit-run parse and never counts as a pre-release.
     */
    fun compare(a: String, b: String): Int {
        val (ca, pa) = split(a)
        val (cb, pb) = split(b)
        val x = parts(ca)
        val y = parts(cb)
        for (i in 0 until maxOf(x.size, y.size)) {
            val u = x.getOrElse(i) { 0L }
            val v = y.getOrElse(i) { 0L }
            if (u != v) return u.compareTo(v)
        }
        if (pa == null && pb == null) return 0
        if (pa == null) return 1
        if (pb == null) return -1
        for (i in 0 until minOf(pa.size, pb.size)) {
            val c = compareIdentifier(pa[i], pb[i])
            if (c != 0) return c
        }
        return pa.size.compareTo(pb.size)
    }

    /** True for development pre-releases (`vX.Y.Z-dev.N`). */
    fun isDev(tag: String): Boolean = norm(tag).lowercase().contains("-dev.")

    /** The version without its pre-release suffix — `v0.1.0-dev.2` -> `0.1.0` (an APK's own versionName). */
    fun core(tag: String): String = split(tag).first

    private fun split(s: String): Pair<String, List<String>?> {
        val n = norm(s)
        val dash = n.indexOf('-')
        if (n.isEmpty() || !n[0].isDigit() || dash < 0) return n to null
        return n.substring(0, dash) to n.substring(dash + 1).split('.')
    }

    private fun compareIdentifier(a: String, b: String): Int {
        val na = a.toLongOrNull()
        val nb = b.toLongOrNull()
        return when {
            na != null && nb != null -> na.compareTo(nb)
            na != null -> -1
            nb != null -> 1
            else -> a.compareTo(b)
        }
    }

    private fun parts(s: String): List<Long> = norm(s).split('.').map { segment ->
        // First contiguous digit run in the segment: skip leading non-digits, take the digits, stop at
        // the next non-digit. Handles date-style tags ("build-20260912" -> 20260912) while staying
        // identical for ordinary semver segments ("2", "0-rc1" -> 0).
        var n = 0L
        var started = false
        for (c in segment) {
            if (c.isDigit()) {
                started = true
                n = if (n > (Long.MAX_VALUE - 9) / 10) Long.MAX_VALUE else n * 10 + (c - '0')
            } else if (started) break
        }
        n
    }
}

